File: apis/server/test/weather_test.js
/* */

const chai = require('chai');
const chaiHttp = require('chai-http');
const chaiXml = require('chai-xml');

const expect = require('chai').expect;

chai.use(chaiHttp);
chai.use(chaiXml);

const url = 'https://api.codetabs.com';
// const url = 'http://localhost:3000'

describe('WEATHER TEST ', function () {
  before(function (done) {
    done();
  });
  if (url !== 'http://localhost:3000') {
    it('local temp // fails on localhost', function (done) {
      chai.request(url)
        .get('/weather/temp')
        .query({})
        .end(function (err, res) {
          // console.log('PATH=> ', res.req.path)
          expect(err).to.be.null;
          expect(res).to.have.status(200);
          expect(res).to.be.json;
          expect(res.body).to.have.property('city');
          expect(res.body).to.have.property('country');
          expect(res.body).to.have.property('tempC');
          expect(res.body).to.have.property('tempF');
          done();
        });
    });
  }
  if (url === 'http://localhost:3000') {
    it('empty city', function (done) {
      chai.request(url)
        .get('/weather/temp?city=')
        .query({})
        .end(function (err, res) {
          // console.log('PATH=> ', res.req.path)
          expect(err).to.be.null;
          expect(res).to.have.status(400);
          done();
        });
    });
  } else {
    it('local temp', function (done) {
      chai.request(url)
        .get('/weather/temp')
        .query({})
        .end(function (err, res) {
          // console.log('PATH=> ', res.req.path)
          expect(err).to.be.null;
          expect(res).to.have.status(200);
          expect(res).to.be.json;
          expect(res.body).to.have.property('city');
          expect(res.body).to.have.property('country');
          expect(res.body).to.have.property('tempC');
          expect(res.body).to.have.property('tempF');
          done();
        });
    });
  }
  it('inexistent city', function (done) {
    chai.request(url)
      .get('/weather/temp?city=inexistentcity')
      .query({})
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(404);
        expect(res.body).to.be.equal('Error 404');

        done();
      });
  });
  it('london city ', function (done) {
    chai.request(url)
      .get('/weather/temp?city=london')
      .query({})
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.have.property('city');
        expect(res.body).to.have.property('country');
        expect(res.body).to.have.property('tempC');
        expect(res.body).to.have.property('tempF');
        done();
      });
  });
  it('new york city', function (done) {
    chai.request(url)
      .get('/weather/temp?city=new york')
      .query({})
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.have.property('city');
        expect(res.body).to.have.property('country');
        expect(res.body).to.have.property('tempC');
        expect(res.body).to.have.property('tempF');
        done();
      });
  });
});

File: weather/weather_test.go
/* */

package weather

import (
	"encoding/json"
	"fmt"
	"net/http"
	"testing"
)

type weatherTestOutput struct {
	StatusCode int
	Error      string  `json:"Error,omitempty"`
	TempC      float64 `json:"tempC"`
	TempF      float64 `json:"tempF"`
	City       string  `json:"city"`
	Country    string  `json:"country"`
	Lat        float64 `json:"latitude"`
	Lon        float64 `json:"longitude"`
}

func TestWeatherApi(t *testing.T) {
	const (
		serverURL   = "http://localhost:3000"
		validFormat = "Bad request, valid format is 'api.codetabs.com/v1/{service}?{param}=value'. Please read our docs at https://codetabs.com"
	)

	var weatherTests = []struct {
		it         string
		endpoint   string
		errorText  string
		statusCode int
	}{
		{"empty", "/v1/weather", "", 406},
		{"bad city", "/v1/weather?city=notRealCity", "", 200},
		{"bad city", "/v1/weather/?city=lon%%don", validFormat, 400},
		{"incomplete", "/v1/weather?city=&format=xml", "", 406},
		{"bad", "/v1/weather/bad", validFormat, 400},
		{"local temp", "/v1/weather/", validFormat, 400},
		{"empty city", "/v1/weather?city=", "", 406},
		{"london", "/v1/weather?city=london", "", 200},
		{"new york", "/v1/weather?city=new+york", "", 200},
	}

	for _, test := range weatherTests {
		var to weatherTestOutput

		req, err := http.NewRequest("GET", fmt.Sprintf("%s%s", serverURL, test.endpoint), nil)
		if err != nil {
			t.Errorf("Error preparing request for %s: %v", test.it, err)
			continue
		}

		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatalf("Error making request for %s: %v", test.it, err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != test.statusCode {
			t.Errorf("%s got %v, expected %v", test.it, resp.StatusCode, test.statusCode)
		}

		err = json.NewDecoder(resp.Body).Decode(&to)
		if err != nil {
			t.Fatalf("Error decoding response for %s: %v", test.it, err)
		}

		if to.Error != test.errorText {
			t.Errorf("%s got '%v', expected '%v'", test.it, to.Error, test.errorText)
		}

		if test.statusCode == 200 && test.errorText == "" {
			if to.City == "" && to.TempC == 0 && to.Lat == 0 && to.Lon == 0 {
				t.Errorf("%s: incomplete or incorrect weather data", test.it)
			}
		}

		fmt.Printf("Test OK... %s\n", test.it)
	}
}
